diff -urN a/binutils-2.23.1/bfd/elfxx-mips.c b/binutils-2.23.1/bfd/elfxx-mips.c
--- a/binutils-2.23.1/bfd/elfxx-mips.c	2012-07-13 12:58:45.723590998 +0100
+++ b/binutils-2.23.1/bfd/elfxx-mips.c	2012-07-13 13:00:14.083338468 +0100
@@ -4026,6 +4026,10 @@
   struct mips_elf_link_hash_table *htab;
   struct mips_got_info *g;
 
+  /* Follow warning symbol link.  */
+  if (h->root.root.type == bfd_link_hash_warning)
+    h = (struct mips_elf_link_hash_entry *) h->root.root.u.i.link;
+
   info = (struct bfd_link_info *) data;
   htab = mips_elf_hash_table (info);
   g = htab->got_info;
@@ -8999,6 +8999,10 @@
 {
   struct mips_elf_link_hash_table *htab;
 
+  /* Follow warning symbol link.  */
+  if (h->root.root.type == bfd_link_hash_warning)
+    h = (struct mips_elf_link_hash_entry *) h->root.root.u.i.link;
+
   htab = (struct mips_elf_link_hash_table *) data;
   if (h->needs_lazy_stub)
     {
